### PR TITLE
Fix possible size_t overflow in 32-bit builds.

### PR DESCRIPTION
**Pull request by JohnKesapidesARM**
Use cl_ulong temporary values for row/slice_pitch.

Signed-off-by: John Kesapides <john.kesapides@arm.com>

### CODE REVIEW

File: test_conformance/images/kernel_image_methods/test_3D.cpp [hunk 1]
```diff
@@ -113,23 +116,26 @@ int test_get_image_info_3D(cl_device_id device, cl_context context,
...
 
                 size_t extraHeight = (int)random_log_in_range( 0, 8, seed );
-                imageInfo.slicePitch = imageInfo.rowPitch * (imageInfo.height + extraHeight);
+                slicePitch = rowPitch * (imageInfo.height + extraHeight);
 
-                size = (cl_ulong)imageInfo.slicePitch * (cl_ulong)imageInfo.depth * 4 * 4;
+                size = (cl_ulong)slicePitch * (cl_ulong)imageInfo.depth * 4 * 4;
             } while(  size > maxAllocSize || ( size * 3 ) > memSize );
 
+            imageInfo.slicePitch = slicePitch;
```
Comment (jlewis-austin) [inline test_conformance/images/kernel_image_methods/test_3D.cpp:136]: I'm not sure I understand the fix. Aren't we just moving the overflow possibility to here? It seems like a fundamental limitation of the API that rowPitch and slicePitch have to fit in size_t.

File: test_conformance/images/kernel_image_methods/test_2D_array.cpp [hunk 2]
```diff
@@ -252,23 +255,27 @@ int test_get_image_info_2D_array(cl_device_id device, cl_context context,
...
-                } while ((imageInfo.rowPitch % pixelSize) != 0);
+                    rowPitch += extraWidth;
+                } while ((rowPitch % pixelSize) != 0);
 
                 size_t extraHeight = (int)random_log_in_range( 0, 8, seed );
-                imageInfo.slicePitch = imageInfo.rowPitch * (imageInfo.height + extraHeight);
+                slicePitch = rowPitch * (imageInfo.height + extraHeight);
 
-                size = (cl_ulong)imageInfo.slicePitch * (cl_ulong)imageInfo.arraySize * 4 * 4;
+                size = (cl_ulong)slicePitch * (cl_ulong)imageInfo.arraySize * 4
```
Comment (jlewis-austin) [inline test_conformance/images/kernel_image_methods/test_2D_array.cpp:272]: the (cl_ulong) cast is unnecessary here

File: test_conformance/images/kernel_image_methods/test_3D.cpp [hunk 3]
```diff
@@ -113,23 +116,26 @@ int test_get_image_info_3D(cl_device_id device, cl_context context,
...
-                } while ((imageInfo.rowPitch % pixelSize) != 0);
+                    rowPitch += extraWidth;
+                } while ((rowPitch % pixelSize) != 0);
 
                 size_t extraHeight = (int)random_log_in_range( 0, 8, seed );
-                imageInfo.slicePitch = imageInfo.rowPitch * (imageInfo.height + extraHeight);
+                slicePitch = rowPitch * (imageInfo.height + extraHeight);
 
-                size = (cl_ulong)imageInfo.slicePitch * (cl_ulong)imageInfo.depth * 4 * 4;
+                size = (cl_ulong)slicePitch * (cl_ulong)imageInfo.depth * 4 * 4;
```
Comment (jlewis-austin) [inline test_conformance/images/kernel_image_methods/test_3D.cpp:133]: the (cl_ulong) cast is unnecessary here

**Review by jlewis-austin — COMMENTED** [review]
This definitely cleans it up a bit, but think I'm missing how the fix works.

File: test_conformance/images/kernel_image_methods/test_3D.cpp (hunk 1 above)
Comment (JohnKesapidesARM) [reply test_conformance/images/kernel_image_methods/test_3D.cpp:136]: The trick here is that for a 32-bit system, if the pitch is more than 32-bit then this `while(  size > maxAllocSize || ( size * 3 ) > memSize );` will cause it to recalculate as maxAllocSize and memSize  are guaranteed to be <=4GB. So if we get to this point outside the loop the pitches are guaranteed to fit.

File: test_conformance/images/kernel_image_methods/test_2D_array.cpp (hunk 2 above)
Comment (JohnKesapidesARM) [reply test_conformance/images/kernel_image_methods/test_2D_array.cpp]: yeap redundant, removed.

File: test_conformance/images/kernel_image_methods/test_3D.cpp (hunk 3 above)
Comment (JohnKesapidesARM) [reply test_conformance/images/kernel_image_methods/test_3D.cpp]: yeap redundant, removed.

File: test_conformance/images/kernel_image_methods/test_3D.cpp (hunk 1 above)
Comment (jlewis-austin) [reply test_conformance/images/kernel_image_methods/test_3D.cpp:136]: Thanks, got it. I was looking at the wrong loop scope.

**Review by jlewis-austin — APPROVED** [review]
LGTM

**Review by kpet — APPROVED** [review]

**Pull request merged**